pallets/foreign-investments/src/impls/benchmark_utils.rs

```diff
@@ -0,0 +1,164 @@
+// Copyright 2021 Centrifuge Foundation (centrifuge.io).
+// This file is part of Centrifuge Chain project.
+
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+use cfg_traits::{
+	benchmarking::{
+		ForeignInvestmentBenchmarkHelper, InvestmentIdBenchmarkHelper, OrderBookBenchmarkHelper,
+		PoolBenchmarkHelper,
+	},
+	investments::{ForeignInvestment, OrderManager},
+};
+use cfg_types::{
+	fixed_point::Ratio,
+	orders::{FulfillmentWithPrice, TotalOrder},
+	tokens::CurrencyId,
+};
+use frame_benchmarking::Zero;
+use frame_support::assert_ok;
+use sp_runtime::{DispatchError, FixedPointNumber, Perquintill};
+
+use crate::{Config, Pallet};
+
+pub const CURRENCY_POOL: CurrencyId = CurrencyId::ForeignAsset(1);
+pub const CURRENCY_FOREIGN: CurrencyId = CurrencyId::ForeignAsset(2);
+pub const DECIMALS_POOL: u32 = 12;
+pub const DECIMALS_FOREIGN: u32 = 6;
+pub const INVEST_AMOUNT_POOL_DENOMINATED: u128 = 1_000_000_000_000;
+pub const INVEST_AMOUNT_FOREIGN_DENOMINATED: u128 = INVEST_AMOUNT_POOL_DENOMINATED / 1_000_000;
+
+impl<T: Config> ForeignInvestmentBenchmarkHelper for Pallet<T>
+where
+	T::Balance: From<u128>,
+	T::CurrencyId: From<CurrencyId>,
+	T::PoolInspect: PoolBenchmarkHelper<PoolId = T::PoolId, AccountId = T::AccountId, Balance = T::Balance>
+		+ InvestmentIdBenchmarkHelper<PoolId = T::PoolId, InvestmentId = T::InvestmentId>,
+	T::TokenSwaps: OrderBookBenchmarkHelper<
+		AccountId = T::AccountId,
+		Balance = T::Balance,
+		CurrencyId = T::CurrencyId,
+		OrderIdNonce = T::TokenSwapOrderId,
+	>,
+	T::Investment: OrderManager<
+		Error = DispatchError,
+		InvestmentId = T::InvestmentId,
+		Orders = TotalOrder<T::Balance>,
+		Fulfillment = FulfillmentWithPrice<T::BalanceRatio>,
+	>,
+	T::BalanceRatio: From<Ratio>,
+{
+	type AccountId = T::AccountId;
+	type Balance = T::Balance;
+	type CurrencyId = T::CurrencyId;
+	type InvestmentId = T::InvestmentId;
+
+	fn bench_prepare_foreign_investments_setup() -> (
+		T::AccountId,
+		T::InvestmentId,
+		T::CurrencyId,
+		T::CurrencyId,
+		T::AccountId,
+	) {
+		let pool_id = Default::default();
+		let pool_admin: T::AccountId = frame_benchmarking::account("pool_admin", 0, 0);
+		<T::PoolInspect as PoolBenchmarkHelper>::bench_create_pool(pool_id, &pool_admin);
```

Comment:
Comment: I really like how the `bench_*` methods are called inside `bench_*` methods, leaving a clean API to use from the benchmark side with 0 knowledge about what's happening under the hood. i.e. the `pool_id` used is totally transparent here ❤️